File: spoj/buglife.py
#!/usr/bin/python
# -*- encoding: utf-8 -*-
# pylint: disable=invalid-name,missing-docstring,bad-builtin
from sys import stdin
from collections import deque
from itertools import imap, izip

def main():
    def is_suspicious(gender, start, interations, status):
        gender[start] = status
        queue = deque([start])
        while queue:
            node = queue.popleft()
            for mate in interations[node]:
                if gender[mate] is None:
                    gender[mate] = not gender[node]
                    queue.append(mate)
                elif gender[mate] == gender[node]:
                    return True  # suspicious bugs found

    dstream = imap(int, stdin.read().split())
    for t in xrange(1, next(dstream) + 1):
        n, i = next(dstream), next(dstream)
        interations = [[] for _ in xrange(n)]
        for _, a, b in izip(xrange(i), dstream, dstream):
            interations[a - 1].append(b - 1)
            interations[b - 1].append(a - 1)

        arr = [None] * n
        print "Scenario #%s:" % t
        for x in xrange(n):
            if arr[x] is None and is_suspicious(arr, x, interations, True):
                print "Suspicious bugs found!"
                break
        else:
            print "No suspicious bugs found!"

main()

File: spoj/buglife.cc
#include <bits/stdc++.h>
using namespace std;
#define usingcincout ios::sync_with_stdio(0);cin.tie(0);

template <class T>
void gint(T& n) {
    n = 0;
    int sign=1;
    register char c=0;
    while(c<33)
        c=getchar_unlocked();
    if (c=='-'){
        sign=-1;
        c=getchar_unlocked();
    }
    while(c>='0'&&c<='9'){
        n=(n<<3)+(n<<1)+(c-'0');
        c=getchar_unlocked();
    } n *= sign;
}
template <class T>
void dint(T a) {
    char num[20];
    int i = 0;
    if(a < 0) {
        putchar_unlocked('-');
        a *= -1;
    }
    do {
        num[i++] = a%10 + 48;
        a /= 10;
    }  while (a != 0);
    i--;
    while (i >= 0)
        putchar_unlocked(num[i--]);
}
void dstr(char* s)
{
    while(*s)
        putchar_unlocked(*(s++));
}
bool is_suspicious(int32_t gender[], int32_t start, vector<int32_t> interations[], int32_t status)
{
    gender[start] = status;
    queue<int32_t> Q;
    Q.push(start);
    while(Q.size())
    {
        int32_t node = Q.front();
        Q.pop();
        for(vector<int32_t>::iterator mate = interations[node].begin(); mate != interations[node].end(); mate++)
        {
            if(gender[*mate] == 0)
            {
                if(gender[node] == 1){
                    gender[*mate] = 2;
                }
                else{
                    gender[*mate] = 1;
                }
                Q.push(*mate);
            }
            else if(gender[*mate] == gender[node]){
                return true;
            }
        }
    }
    return false;
}
int main(int argc, char *argv[])
{
    int32_t t;
    gint(t);
    for(int32_t _ = 0; _ < t; _++)
    {
        int32_t n, i;
        gint(n);
        gint(i);
        vector<int32_t> interations[n];
        for (int32_t x = 0; x < i; ++x)
        {
            int32_t a, b;
            gint(a);
            gint(b);
            interations[a - 1].push_back(b - 1);
            interations[b - 1].push_back(a - 1);
        }
        int32_t arr[n];
        memset(arr, 0, sizeof(int32_t)*n);
        dstr("Scenario #");
        dint(_ + 1);
        putchar_unlocked(':');
        putchar_unlocked('\n');
        bool flag = true;
        for(int32_t x = 0; x < n; x++)
        {
            if((arr[x] == 0) and (is_suspicious(arr, x, interations, 1)))
            {
                dstr("Suspicious bugs found!\n");
                flag = false;
                break;
            }
        }
        if(flag)
            dstr("No suspicious bugs found!\n");
    }
    return 0;
}
